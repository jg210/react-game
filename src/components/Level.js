// @flow
import React, { Component } from 'react';
import { connect } from "react-redux";
import _ from 'lodash';

import { levelChange } from '../redux/actions';
import type { LevelState } from '../redux/reducers/level';

type Props = {
  level: number,
  numberOfLevels: number,
  levelChange: (number) => void
}

class Level extends Component<Props> {

  render() {
    return <select value={this.props.level} onChange={this.handleClick}> 
      {_.range(1, this.props.numberOfLevels + 1).map((i: number) =>
        <option key={i} value={i}>{i}</option>
      )}
    </select>;
  }

  handleClick = (event: SyntheticEvent<HTMLSelectElement>) => {
    const target: HTMLSelectElement = event.currentTarget;
    if (target) {
      const level: number = parseInt(target.value);
      this.props.levelChange(level);
    }
  }

}

const mapStateToProps = (state: {level: LevelState}) => {
  const level = state.level.current;
  const numberOfLevels = state.level.last
  return { level, numberOfLevels };
};
const actionCreators = {
  levelChange
};
export default connect(mapStateToProps, actionCreators)(Level);
