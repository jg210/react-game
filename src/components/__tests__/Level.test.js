// @flow
//
// (c) 2019 Jeremy Green

import React from 'react';
import { Level } from '../Level';

import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

it('renders correctly', () => {
  const tree = renderer
    .create(<Level
      level={3}
      numberOfLevels={5}
      levelChange={() => {}}
    />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('sends event when clicked on', () => {
  const numberOfLevels = 3;
  for (let level = 1 ; level <= numberOfLevels; level++) {
    for (let initialLevel = 1 ; initialLevel <= numberOfLevels; initialLevel++) {
      const handler = jest.fn();
      const component = mount(<Level
        level={initialLevel}
        numberOfLevels={numberOfLevels}
        levelChange={handler}/>
      );
      const select = component.find('select');
      const selectDOMNode = select.getDOMNode();
      selectDOMNode.value = level;
      select.simulate('change', { target: selectDOMNode });
      expect(handler).toHaveBeenCalledTimes(1);
      expect(handler).toHaveBeenCalledWith(level);
    }
  }
});