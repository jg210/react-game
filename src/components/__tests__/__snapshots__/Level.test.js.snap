// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<select
  onChange={[Function]}
  value={3}
>
  <option
    value={1}
  >
    1
  </option>
  <option
    value={2}
  >
    2
  </option>
  <option
    value={3}
  >
    3
  </option>
  <option
    value={4}
  >
    4
  </option>
  <option
    value={5}
  >
    5
  </option>
</select>
`;
